### PR TITLE
Support arbitrary shaped Material.

### PR DESCRIPTION
**Pull request by amirh**
For backward compatibility we keep supporting specifying the shape as a
combination of MaterialType and borderRadius, and we just use that as a
default when shapeBorder is null.

To cleanup the implementation if shapeBorder was not specified we just
translate the specified shape to a shapeBorder internally.
I benchmarked paint, layout and hit testing, with the specialized shape
clippers vs. the equivalent path clippers and did not see any
significant performance difference.

For testing, I extended the clippers/physicalShape matchers to match either the
specialized shape or the equivalent shape specified with a path.

### CODE REVIEW

File: packages/flutter/lib/src/material/material.dart [hunk 1]
```diff
@@ -277,41 +281,60 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+      child: contents,
+      clipper: new ShapeBorderClipper(
+        shapeBorder: shapeBorder,
+      ),
+    );
+  }
+
+  // Determines the shape for this Material.
+  //
+  // If a shapeBorder was specified it will determine the shape,
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: specified --> specified,
shape, --> shape.

File: packages/flutter/lib/src/material/material.dart [hunk 2]
```diff
@@ -277,41 +281,60 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+  // Otherwise if a borderRadius was specified, the shape is a rounded
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: Otherwise if --> If

File: packages/flutter/lib/src/material/material.dart [hunk 3]
```diff
@@ -277,41 +281,60 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+  // rectangle.
+  // Otherwise the shape is determined by the widget type as described in the
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: Otherwise --> Otherwise,

File: packages/flutter/lib/src/material/material.dart [hunk 4]
```diff
@@ -475,3 +498,134 @@ abstract class InkFeature {
   @override
   String toString() => describeIdentity(this);
 }
+
+/// An interpolation between to [ShapeBorder]s.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: to --> two

File: packages/flutter/lib/src/material/material.dart [hunk 5]
```diff
@@ -475,3 +498,134 @@ abstract class InkFeature {
...
+///
+/// This class specialises the interpolation of [Tween] to use [ShapeBorder.lerp].
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: specialises --> specializes

File: packages/flutter/lib/src/material/material.dart [hunk 6]
```diff
@@ -475,3 +498,134 @@ abstract class InkFeature {
...
+class ShapeBorderTween extends Tween<ShapeBorder> {
+
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: Don't need this extra blank line.

File: packages/flutter/lib/src/material/material.dart [hunk 7]
```diff
@@ -475,3 +498,134 @@ abstract class InkFeature {
...
+  /// the [begin] and [end] properties may be null; see [ShapeBorder.lerp] for
+  /// the null handling semantics.
+  ShapeBorderTween({ShapeBorder begin, ShapeBorder end}): super(begin: begin, end: end);
+
+  /// Returns the value this variable has at the given animation clock value.
+  @override
+  ShapeBorder lerp(double t) {
+    return ShapeBorder.lerp(begin, end, t);
+  }
+
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: Or this blank line.

File: packages/flutter/lib/src/material/material.dart [hunk 8]
```diff
@@ -475,3 +498,134 @@ abstract class InkFeature {
...
+/// This class specialises the interpolation of [Tween] to use [ShapeBorder.lerp].
+class ShapeBorderTween extends Tween<ShapeBorder> {
+
+  /// Creates a [ShapeBorder] tween.
+  ///
+  /// the [begin] and [end] properties may be null; see [ShapeBorder.lerp] for
+  /// the null handling semantics.
+  ShapeBorderTween({ShapeBorder begin, ShapeBorder end}): super(begin: begin, end: end);
+
+  /// Returns the value this variable has at the given animation clock value.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/material.dart]: Not sure what you mean by "variable."  Seems more like it should be "[Tween]".

File: packages/flutter/lib/src/material/material.dart (hunk 2 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 4 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 5 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 6 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 8 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 7 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart [hunk 9]
```diff
@@ -153,6 +155,7 @@ class Material extends StatefulWidget {
     this.shadowColor: const Color(0xFF000000),
     this.textStyle,
     this.borderRadius,
+    this.shapeBorder,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: let's just call this "shape" for consistency with other places where we take ShapeBorder

File: packages/flutter/lib/src/material/material.dart [hunk 10]
```diff
@@ -277,41 +281,60 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
...
+      elevation: widget.elevation,
+      color: backgroundColor,
+      shadowColor: widget.shadowColor,
+      animateColor: false,
+      child: contents,
+    );
+
+  }
+
+  Widget _clipToShape({ShapeBorder shapeBorder, Widget contents}) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: this can be static

File: packages/flutter/lib/src/material/material.dart [hunk 11]
```diff
@@ -475,3 +498,132 @@ abstract class InkFeature {
...
+/// Animated version of [PhysicalShape].
+///
+/// The [elevation] is animated.
+///
+/// The [color] is animated if the [animateColor] property is set; otherwise,
+/// the color changes immediately at the start of the animation for the other
+/// two properties. This allows the color to be animated independently (e.g.
+/// because it is being driven by an [AnimatedTheme]).
+///
+/// The [clipper] is not animated.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: these docs are out of date (e.g. clipper doesn't exist)

File: packages/flutter/lib/src/material/material.dart [hunk 12]
```diff
@@ -475,3 +498,132 @@ abstract class InkFeature {
...
+class _AnimatedPhysicalShape extends ImplicitlyAnimatedWidget {
+  /// Creates a widget that animates the properties of a [PhysicalShape] not
+  /// including the [clipper].
+  ///
+  /// The [child], [clipper], [elevation], [color], [shadowColor], [curve], and
+  /// [duration] arguments must not be null.
+  ///
+  /// Animating [color] is optional and is controlled by the [animateColor] flag.
+  ///
+  /// Animating [shadowColor] is optional and is controlled by the [animateShadowColor] flag.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: ditto

File: packages/flutter/lib/src/material/material.dart [hunk 13]
```diff
@@ -475,3 +498,132 @@ abstract class InkFeature {
...
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The border of the widget.
+  ///
+  /// This border will be painted, and in addition the outer path of the border
+  /// determines the physical shape.
+  final ShapeBorder shapeBorder;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: shape

File: packages/flutter/lib/src/material/material.dart [hunk 14]
```diff
@@ -475,3 +498,132 @@ abstract class InkFeature {
...
+  /// Animating [shadowColor] is optional and is controlled by the [animateShadowColor] flag.
+  const _AnimatedPhysicalShape({
+    Key key,
+    @required this.child,
+    @required this.shapeBorder,
+    @required this.elevation,
+    @required this.color,
+    this.animateColor: true,
+    @required this.shadowColor,
+    this.animateShadowColor: true,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: it seems you can remove animateColor and animateShadowColor

File: packages/flutter/lib/src/material/material.dart [hunk 15]
```diff
@@ -475,3 +498,132 @@ abstract class InkFeature {
...
+  /// Creates a widget that animates the properties of a [PhysicalShape] not
+  /// including the [clipper].
+  ///
+  /// The [child], [clipper], [elevation], [color], [shadowColor], [curve], and
+  /// [duration] arguments must not be null.
+  ///
+  /// Animating [color] is optional and is controlled by the [animateColor] flag.
+  ///
+  /// Animating [shadowColor] is optional and is controlled by the [animateShadowColor] flag.
+  const _AnimatedPhysicalShape({
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: i'd probably call this _MaterialInterior to avoid confusion with PhysicalShape

File: packages/flutter/lib/src/material/material.dart (hunk 14 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/material.dart]: Should they really default to true?

Comment (Hixie) [reply packages/flutter/lib/src/material/material.dart]: oh, i see. I would remove them, but make sure you remove the "true" branches since it's hard-coded to false above.

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/material/material.dart (hunk 9 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done, to be completely consistent I also renamed shapeBorder to shape in ShapeBorderClipper

File: packages/flutter/lib/src/material/material.dart (hunk 10 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart (hunk 14 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: removed, I only removed the true branch for animateColor as animateShadowColor was defaulting to true. It feels a bit weird, but I keep the same behavior as before.

File: packages/flutter/lib/src/material/material.dart (hunk 15 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart [hunk 16]
```diff
@@ -147,6 +147,10 @@ class Material extends StatefulWidget {
   /// Creates a piece of material.
   ///
   /// The [type], [elevation] and [shadowColor] arguments must not be null.
+  ///
+  /// [shape] and [borderRadius] must not be both non-null.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: Avoid starting with a lowercase letter. "The [shape] ... arguments..."

File: packages/flutter/lib/src/material/material.dart [hunk 17]
```diff
@@ -147,6 +147,10 @@ class Material extends StatefulWidget {
...
+  /// If [type] is [MaterialType.circle] then [borderRadius] and [shape] must be
+  /// null.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: Another way to phrase these two additions would be "If a [shape] is specified, then the [borderRadius] property must not be null and the [type] property must not be [MaterialType.circle]. If the [borderRadius] is specified, then the [type] property must not be [MaterialType.circle]. In both cases, these restrictions are intended to catch likely errors."

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/material/material.dart (hunk 17 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

File: packages/flutter/lib/src/material/material.dart [hunk 18]
```diff
@@ -114,9 +114,11 @@ abstract class MaterialInkController {
 ///
```
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart]: maybe rephrase the last paragraph above the Shape header to say that elevation and shadowColor are animated, and not the others.

**Comment by Hixie** [comment]
Is the plan to paint the border in a subsequent patch? (That's fine, just curious.)

File: packages/flutter/lib/src/material/material.dart (hunk 18 above)
Comment (amirh) [reply packages/flutter/lib/src/material/material.dart]: done

**Comment by Hixie** [comment]
This regressed complex_layout_scroll_perf__timeline_summary average_frame_rasterizer_time_millis by a significant amount.

**Comment by Hixie** [comment]
Also complex_layout_scroll_perf__timeline_summary missed_frame_rasterizer_budget_count.
